### ISSUE TITLE
[Go] - use full url on package imports

### ISSUE THREAD

**Issue by mishudark** [thought-machine/please#439]
I see here https://github.com/thought-machine/please/blob/master/src/please.go#L27 the import of package `hashes` instead of `github.com/thought-machine/src/hashes` as usual on Go

is there any way of importing the package using the full URL?

**Comment by peterebden** [comment]
I'm not sure we'd gain much; to import them via a full path we'd need to `go get` the repo, but it's not really go-gettable, there are several build steps that can't be run through `go build`.

It might be possible to do something to work around it, what would it improve if we did?

**Comment by mishudark** [comment]
The main value is to follow the standard go imports rules

**Comment by peterebden** [comment]
Sure, I'm happy to look if someone wants to send a pull request. I'm a little short on time now unfortunately - day job taking up too much time!

**Comment by sean-** [comment]
ha!  I actually just ran into this problem as well and was searching for this.  In our existing repository, we use a path like:

```go
import (
  "os"
  "path/filespec"

  "github.com/myorgname/my-project/foo"
  "github.com/myorgname/my-project/foo/baz"
  ...
)
```

When switching to `plz`, I've noticed this doesn't work and I'll get error response:

```
pkg/config/configdir_config.go, line 9, column 2: can't find import: "github.com/myorgname/my-project/foo"
```

Which is frustrating, because it's there, and installed:

```
./plz-out/go/pkg/darwin_amd64/foo
./plz-out/go/src/pkg/foo
./plz-out/gen/pkg/foo
./plz-out/tmp/pkg/foo
./plz-out/tmp/pkg/config/config._build/pkg/foo
./.plz-cache/pkg/foo
./.plz-cache/pkg/foo/foo
./pkg/foo
```

but not in the location that I'd expect relative to our `GOPATH`.  When I use `go list ./pkg/foo` it gives me back the canonical import path.

The source of friction is, if I want to move to `plz`, I need to change our `import` path and rewrite our headers.  And sacrifice making our binaries `go get`-able.  This, for instance, works:

```go
import (
  "os"
  "path/filespec"

  "foo"
  "foo/baz"
  ...
)
```

?  Is it really necessary to move everything to `go_get()` rules for the same repo we're working in?  This seems odd to me.  I'd be happy with having `go_library()` having an `exported_name` or something like that, which would default to the path returned from `go list`'s canonical understanding of its import path.

?

**Comment by sean-** [comment]
Btw, the reason this is a sticking point is because a `go get`-able application is one that works out of the box with `goimports`, which is a huge boon for developer productivity on the project.  Am I missing something?  We’ve checked out our monorepo into `$GOPATH/github.com/my-orgname/my-monorepo` but there’s a conflict between `goimports` and `plz build` because it looks like `go_library()` doesn’t have or use the canonical path relative to the `$GOPATH` for a user.  Is there a flag that can be passed to `go_library()` to get the desired behavior?

**Comment by peterebden** [comment]
I understand the problem now - I thought originally this referred specifically to this repository.

Right now there is no solution for it on `go_library` or a similar config flag, but ideally there should be. We've not needed it so far since our internal monorepo doesn't have a canonical global GOPATH (it is not hosted on Github). I'm keen on the idea, just not sure what the implementation would look like.

**Comment by peterebden** [comment]
Um, so this is a little embarrassing... turns out we actually do have this feature. I implemented it last year in #249 . Blame advancing age...

If you set the following in your .plzconfig file it should be able to find imports from that path within the local repo:
```
[go]
importpath = github.com/thought-machine/please
```

Hope that helps!

**Comment by sean-** [comment]
I submitted a poorly worded doc PR that hopefully addresses this issue.  Thank you for having already done the work!  :-D
